### PR TITLE
Darwin framework: add replacement subscribe method that includes event report

### LINKED ISSUE #16691: Darwin: synchronous-read API part 2: support generic events in subscription, and expected value cache expiration
#### Problem
Current CHIPDevice subscribeWitthQueue: API only supports known standard attributes. For manufacturer specific extensions or unimplemented standard attributes, subscribeAttributeWithEndpointId:... API can be used. However, in such a case, it forces both of them to set keepPreviousSubscriptions option in order not to cancel each other, limiting the flexibility of application to clear up previous subscriptions.

 #### Proposed Solution
Add a generic report handler argument of CHIPDeviceResponseHandler type to subscribeWithQueue API in addition to the existing standard attribute report handler. Note that if standard event handler argument is added to the API, the same generic report handler can be used to receive any event as well as any attribute.

### PR DESCRIPTION
**Pull request by jtung-apple**
#### Problem
* Darwin framework CHIPDevice was missing event subscription. This PR aims to add it
Fixes #16691

#### Change overview
* zapt templates to generate Obj-C objects corresponding to C++ event structs
* zapt template to populate these objects from events TLVReader
* new CHIPDevice subscribeWithQueue method with `eventReporterHandler` and logic to hook up the handler with the callback mechanism

#### Testing
Work-in-progress - will update testing details later.

### CODE REVIEW

**Comment by jtung-apple** [comment]
Tested this with chip-tool-darwin and locally modified Commands.h. I'll make a separate change to add a test to chip-tool-darwin.

File: src/darwin/Framework/CHIP/CHIPDevice.mm [hunk 1]
```diff
@@ -1291,21 +1356,88 @@ - (instancetype)initWithPath:(const ConcreteDataAttributePath &)path value:(null
...
+    id _Nullable value = nil;
+    NSError * _Nullable error = nil;
+    if (apStatus != nullptr) {
+        error = [CHIPError errorForIMStatus:*apStatus];
+    } else if (apData == nullptr) {
+        error = [CHIPError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+    } else {
+        CHIP_ERROR err;
+        value = CHIPDecodeEventPayload(aEventHeader.mPath, *apData, &err);
+        if (err == CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPDevice.mm:1416]:
```suggestion
        if (err == CHIP_ERROR_IM_MALFORMED_EVENT_PATH) {
```

File: src/darwin/Framework/CHIP/templates/CHIPStructsObjc-src.zapt [hunk 2]
```diff
@@ -19,6 +19,23 @@ NS_ASSUME_NONNULL_BEGIN
 @end
 
 {{/zcl_structs}}
+
+{{#zcl_events}}
+@implementation CHIP{{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase name}}Event
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _priorityLevel = @(0);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/CHIPStructsObjc-src.zapt:28]: Where do we use `priorityLevel`?  It looks unused at first glance...

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/CHIPDevice.h [hunk 3]
```diff
@@ -131,6 +131,15 @@ extern NSString * const kCHIPArrayValueType;
               reportHandler:(void (^)(NSArray * _Nullable value, NSError * _Nullable error))reportHandler
     subscriptionEstablished:(nullable void (^)(void))subscriptionEstablishedHandler;
 
+- (void)subscribeWithQueue:(dispatch_queue_t)queue
+                minInterval:(uint16_t)minInterval
+                maxInterval:(uint16_t)maxInterval
+                     params:(nullable CHIPSubscribeParams *)params
+             cacheContainer:(CHIPAttributeCacheContainer * _Nullable)attributeCacheContainer
+     attributeReportHandler:(void (^)(NSArray * _Nullable value, NSError * _Nullable error))attributeReportHandler
+         eventReportHandler:(nullable void (^)(NSArray * _Nullable value, NSError * _Nullable error))eventReportHandler
```
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/CHIPDevice.h:140]: Great point. I'll add a separate error report handler in this PR - better to have client migrate just one time than having to migrate again later.

File: src/darwin/Framework/CHIP/CHIPDevice.mm [hunk 4]
```diff
@@ -365,12 +409,13 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
     } else {
         // SendAutoResubscribeRequest cleans up the params, even on failure.
         attributePath.release();
+        eventPath.release();
```
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/CHIPDevice.mm:412]: Going to make eventReportHandler nonnull, along with the other change.

File: src/darwin/Framework/CHIP/templates/CHIPEventTLVValueDecoder-src.zapt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+{
+    switch (aPath.mClusterId)
+    {
+    {{#chip_client_clusters}}
+    case Clusters::{{asUpperCamelCase name}}::Id: {
+        using namespace Clusters::{{asUpperCamelCase name}};
+        switch (aPath.mEventId)
+        {
+        {{#zcl_events}}
+        // JEFF(zcl_events) name: {{name}}  parent: {{parent.name}}  type: {{type}}
```
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/templates/CHIPEventTLVValueDecoder-src.zapt:30]: Thanks - I should have removed this before I pushed!

File: src/darwin/Framework/CHIP/templates/CHIPStructsObjc-src.zapt (hunk 2 above)
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/templates/CHIPStructsObjc-src.zapt:28]: Left over from development - will remove.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/CHIPDevice.h [hunk 6]
```diff
@@ -131,6 +131,47 @@ extern NSString * const kCHIPArrayValueType;
...
+ * subscription is established.  This will be _after_ the first (priming) call
+ * to reportHandler.  Note that if the CHIPSubscribeParams are set to
+ * automatically resubscribe this can end up being called more than once.
+ */
+- (void)subscribeWithQueue:(dispatch_queue_t)queue
+                minInterval:(uint16_t)minInterval
+                maxInterval:(uint16_t)maxInterval
+                     params:(nullable CHIPSubscribeParams *)params
+             cacheContainer:(CHIPAttributeCacheContainer * _Nullable)attributeCacheContainer
+     attributeReportHandler:(void (^)(NSArray * _Nullable value))attributeReportHandler
```
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/CHIPDevice.h:170]: Oh yes, now that we've separated them they will all be nonnull. Thanks for catching this. I even wrote about this in the header comment block!

File: src/darwin/Framework/CHIP/CHIPDevice.h [hunk 7]
```diff
@@ -131,6 +131,47 @@ extern NSString * const kCHIPArrayValueType;
...
+ * reported via those objects.
+ *
+ * errorHandler will be called any time there is an error for the
+ * entire subscription (with a non-nil "error"), and terminate the subscription.
+ *
+ * Both report handlers are not supported over XPC at the moment.
+ *
+ * subscriptionEstablished block, if not nil, will be called once the
+ * subscription is established.  This will be _after_ the first (priming) call
+ * to reportHandler.  Note that if the CHIPSubscribeParams are set to
```
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/CHIPDevice.h:162]: Both - will change it to "to both report handlers".

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
Fast-tracking platform-specific code.

**Pull request merged**